core/trino-main/src/main/java/io/trino/sql/planner/EqualityInference.java

```diff
@@ -203,18 +202,17 @@ public static EqualityInference newInstance(Metadata metadata, Expression... exp
     public static EqualityInference newInstance(Metadata metadata, Collection<Expression> expressions)
     {
         DisjointSet<Expression> equalities = new DisjointSet<>();
-        List<Expression> candidates = expressions.stream()
+        Stream<Expression> candidates = expressions.stream()
                 .flatMap(expression -> extractConjuncts(expression).stream())
-                .filter(expression -> isInferenceCandidate(metadata, expression))
-                .collect(Collectors.toList());
+                .filter(expression -> isInferenceCandidate(metadata, expression));
 
-        for (Expression expression : candidates) {
+        candidates.forEach(expression -> {
```

Comment:
inline `candidates` variable, similarily elsewhere 

can you move "Avoid collecting expressions to lists" commit to a new PR?